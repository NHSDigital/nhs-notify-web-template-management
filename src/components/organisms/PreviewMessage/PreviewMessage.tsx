'use client';

import { Details } from 'nhsuk-react-components';
import { PreviewMessageProps } from './PreviewMessage.types';
import { ZodErrorSummary } from '@molecules/ZodErrorSummary/ZodErrorSummary';
import { NHSNotifyRadioButtonForm } from '@molecules/NHSNotifyRadioButtonForm/NHSNotifyRadioButtonForm';

export function PreviewMessage(props: PreviewMessageProps) {
  const { form } = props;
  return (
    <>
      <ZodErrorSummary errorHeading={form.errorHeading} state={form.state} />
      <h1 data-testid='preview-message__heading'>
        <span
          data-testid='preview-message__heading-caption'
          className='nhsuk-caption-l'
        >
          {props.sectionHeading}
        </span>
        {props.templateName}
      </h1>
      <Details>
        <Details.Summary data-testid='preview-message-details__heading'>
          {props.details.heading}
        </Details.Summary>
        <Details.Text data-testid='preview-message-details__text'>
          {props.details.text.map((val, idx) => (
            <p key={`details-text-${idx}`}>{val}</p>
          ))}
        </Details.Text>
      </Details>
      {props.PreviewComponent}
      <NHSNotifyRadioButtonForm
        radiosId={form.radiosId}
        action={form.action}
        state={form.state}
        pageHeading={form.pageHeading}
        options={form.options}
        buttonText={form.buttonText}
        legend={form.legend}
      />
    </>
  );
}
