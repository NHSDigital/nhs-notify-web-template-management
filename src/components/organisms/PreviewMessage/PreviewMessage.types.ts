import { Preview } from '@molecules/Preview/Preview';
import { NHSNotifyRadioButtonFormProps } from '@molecules/NHSNotifyRadioButtonForm/NHSNotifyRadioButtonForm';

export type PreviewMessageProps = {
  sectionHeading: string;
  templateName: string;
  details: {
    heading: string;
    text: string[];
  };
  form: {
    errorHeading: string;
  } & NHSNotifyRadioButtonFormProps;
  PreviewComponent: React.ReactElement<typeof Preview>;
};
