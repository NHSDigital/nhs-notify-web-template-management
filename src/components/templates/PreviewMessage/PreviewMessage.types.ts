import { Radios } from 'nhsuk-react-components';
import { Preview } from '../../molecules/Preview/Preview';

export const PREVIEW_TYPES = [
  'Email',
  'NHS app message',
  'Text message',
  'Letter',
] as const;

export type PreviewType = (typeof PREVIEW_TYPES)[number];

export type PreviewMessageProps = {
  type: PreviewType;
  templateName: string;
  details: {
    heading: string;
    text: string[];
  };
  PreviewComponent: React.ReactElement<typeof Preview>;
  FormOptionsComponent: React.ReactElement<typeof Radios>;
};
