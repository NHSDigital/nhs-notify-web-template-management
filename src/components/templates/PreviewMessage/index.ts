export * from './PreviewMessage';
export * from './PreviewMessage.types';
