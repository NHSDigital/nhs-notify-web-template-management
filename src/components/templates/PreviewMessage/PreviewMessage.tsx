'use client';

import { Button, Details, Fieldset, Form } from 'nhsuk-react-components';
import { PreviewMessageProps } from './PreviewMessage.types';

export function PreviewMessage(props: PreviewMessageProps) {
  return (
    <>
      <h1 data-testid='preview-message__heading'>
        <span
          data-testid='preview-message__heading-caption'
          className='nhsuk-caption-l'
        >
          {props.type} template
        </span>
        {props.templateName}
      </h1>
      <Details>
        <Details.Summary data-testid='preview-message-details__heading'>
          {props.details.heading}
        </Details.Summary>
        <Details.Text data-testid='preview-message-details__text'>
          {props.details.text.map((val, idx) => (
            <p key={`details-text-${idx}`}>{val}</p>
          ))}
        </Details.Text>
      </Details>
      {props.PreviewComponent}
      <Form>
        <Fieldset>
          <Fieldset.Legend data-testid='preview-message-form__legend' size='m'>
            What would you like to do next with this template?
          </Fieldset.Legend>
          {props.FormOptionsComponent}
        </Fieldset>
        <Button data-testid='preview-message-form__button'>Continue</Button>
      </Form>
    </>
  );
}
