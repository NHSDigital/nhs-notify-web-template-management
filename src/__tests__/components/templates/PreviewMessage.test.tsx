import { render, screen } from '@testing-library/react';
import { PreviewMessage, PREVIEW_TYPES } from '@templates/PreviewMessage';

describe('PreviewMessage component', () => {
  it('matches snapshot', () => {
    const container = render(
      <PreviewMessage
        type='NHS app message'
        templateName='Example NHS APP template'
        details={{ heading: 'Details heading', text: ['Details text'] }}
        PreviewComponent={<>Preview</>}
        FormOptionsComponent={<>Form</>}
      />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('renders component correctly', () => {
    render(
      <PreviewMessage
        type='Email'
        templateName='Example template'
        details={{ heading: 'Details heading', text: ['Details text'] }}
        PreviewComponent={<>Preview</>}
        FormOptionsComponent={<>Form</>}
      />
    );

    expect(
      screen.getByTestId('preview-message__heading-caption')
    ).toHaveTextContent('Email template');

    expect(screen.getByTestId('preview-message__heading')).toHaveTextContent(
      'Example template'
    );

    expect(
      screen.getByTestId('preview-message-details__heading')
    ).toHaveTextContent('Details heading');

    expect(
      screen.getByTestId('preview-message-details__text')
    ).toHaveTextContent('Details text');

    expect(
      screen.getByTestId('preview-message-form__legend')
    ).toHaveTextContent('What would you like to do next with this template?');

    expect(
      screen.getByTestId('preview-message-form__button')
    ).toHaveTextContent('Continue');
  });

  it.each(PREVIEW_TYPES)('renders type %s correctly', (type) => {
    render(
      <PreviewMessage
        type={type}
        templateName='Example template'
        details={{ heading: 'Details heading', text: ['Details text'] }}
        PreviewComponent={<>Preview</>}
        FormOptionsComponent={<>Form</>}
      />
    );

    expect(
      screen.getByTestId('preview-message__heading-caption')
    ).toHaveTextContent(`${type} template`);
  });
});
