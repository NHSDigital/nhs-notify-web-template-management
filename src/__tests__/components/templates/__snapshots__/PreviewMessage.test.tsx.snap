// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewMessage component matches snapshot 1`] = `
<DocumentFragment>
  <h1
    data-testid="preview-message__heading"
  >
    <span
      class="nhsuk-caption-l"
      data-testid="preview-message__heading-caption"
    >
      NHS app message template
    </span>
    Example NHS APP template
  </h1>
  <details
    class="nhsuk-details"
  >
    <summary
      class="nhsuk-details__summary"
      data-testid="preview-message-details__heading"
    >
      <span
        class="nhsuk-details__summary-text"
      >
        Details heading
      </span>
    </summary>
    <div
      class="nhsuk-details__text"
      data-testid="preview-message-details__text"
    >
      <p>
        Details text
      </p>
    </div>
  </details>
  Preview
  <form>
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m"
        data-testid="preview-message-form__legend"
      >
        What would you like to do next with this template?
      </legend>
      Form
    </fieldset>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="preview-message-form__button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;
